src/category_theory/cofiltered_system.lean

```diff
@@ -215,10 +293,81 @@ let ⟨i, f, h⟩ := F.is_mittag_leffler_iff_eventual_range.1 h j in
 by { rw [to_eventual_ranges_obj, h], apply_instance }
 
 /-- If `F` has all arrows surjective, then it "factors through a poset". -/
-lemma thin_diagram_of_surjective (Fsur : ∀ (i j : J) (f : i ⟶ j), (F.map f).surjective)
-  (i j) (f g : i ⟶ j) : F.map f = F.map g :=
+lemma thin_diagram_of_surjective (Fsur : ∀ ⦃i j : J⦄ (f : i ⟶ j), (F.map f).surjective)
+  {i j} (f g : i ⟶ j) : F.map f = F.map g :=
 let ⟨k, φ, hφ⟩ := cone_maps f g in
-(Fsur k i φ).injective_comp_right $ by simp_rw [← types_comp, ← F.map_comp, hφ]
+(Fsur φ).injective_comp_right $ by simp_rw [← types_comp, ← F.map_comp, hφ]
+
+lemma to_preimages_nonempty_of_surjective [hFn : ∀ (j : J), nonempty (F.obj j)]
+  (Fsur : ∀ ⦃i j : J⦄ (f : i ⟶ j), (F.map f).surjective)
+  (hs : s.nonempty) : ∀ j, nonempty ((F.to_preimages s).obj j) := λ j, by
+begin
+  simp only [to_preimages_obj, nonempty_coe_sort, nonempty_Inter, mem_preimage],
+  obtain (h|⟨⟨ji⟩⟩) := is_empty_or_nonempty (j ⟶ i),
+  { exact ⟨(hFn j).some, λ ji, h.elim ji⟩, },
+  { obtain ⟨y,ys⟩ := hs,
+    obtain ⟨x,rfl⟩ := Fsur ji y,
+    exact ⟨x, λ ji', (F.thin_diagram_of_surjective Fsur ji' ji).symm ▸ ys⟩, },
+end
+
+
+lemma eval_section_injective_of_eventually_injective
+  {j} (Finj : ∀ i (f : i ⟶ j), (F.map f).injective) :
+   ∀ i (f : i ⟶ j), (λ s : F.sections, s.val j).injective :=
+begin
+  refine λ i f s₀ s₁ h, subtype.ext $ funext $ λ k, _,
+  obtain ⟨m, mi, mk, _⟩ := cone_objs i k,
+  dsimp at h,
+  rw [←s₀.prop (mi ≫ f), ←s₁.prop (mi ≫ f)] at h,
+  rw [←s₀.prop mk, ←s₁.prop mk],
+  refine congr_arg _ (Finj m (mi ≫ f) h),
+end
+
+section fintype_cofiltered_system
+
+variables [∀ (j : J), nonempty (F.obj j)] [∀ (j : J), fintype (F.obj j)]
+  (Fsur : ∀ ⦃i j : J⦄ (f :i ⟶ j), (F.map f).surjective)
+
+include Fsur
+lemma eval_section_surjective_of_surjective (i : J) :
+  (λ s : F.sections, s.val i).surjective := λ x, by
```

Comment:
Here `fintype` should also be replaced by `finite` since the lemma doesn't need `fintype` to be stated. The linter doesn't fire however, probably because this is a pi instance?